draft-ietf-dnsop-svcb-https.md

```diff
@@ -1082,18 +1082,37 @@ single-client granularity.
 
 ## Interaction with Alt-Svc
 
-Clients that do not implement support for Encrypted ClientHello MAY
-skip the HTTPS RR query
-if a usable Alt-Svc value is available in the local cache.
-If Alt-Svc connection fails, these clients SHOULD fall back to the HTTPS RR
-client connection procedure ({{client-behavior}}).
-
-Clients that implement support for ECH MUST perform the HTTPS RR query first,
-and MUST only make use of Alt-Svc when operating in SVCB-optional mode (see
-{{ech-client-behavior}}).
-
-This specification does not alter the DNS records used when connecting
-to an Alt-Svc hostname (typically A and/or AAAA only).
+Clients that implement support for both Alt-Svc and HTTPS records SHOULD
+retrieve any HTTPS records for the Alt-Svc alt-authority, and ensure that
+their connection attempts are consistent with both the Alt-Svc parameters
+and any received HTTPS SvcParams.  If present, the HTTPS record's target
+and port override the alt-authority.  For example, suppose that
+"https://example.com" sends an Alt-Svc field value of:
+
+~~~ HTTP
+Alt-Svc: h2="alt.example.com:443", h3=":8443"
+~~~
+
+The client would retrieve the following HTTPS records:
+
+    alt.example.com. IN HTTPS 1 . alpn=h2,h3 ech=...
+    _8443._https.example.com. IN HTTPS 1 alt3.example.com. (
+        port=9443 alpn=h2,h3 ech=... )
+
+The client could then attempt an HTTP/3 connection to `alt3.example.com:9443`
+with ECH, or an HTTP over TLS connection to `alt.example.com:443` with ECH,
+as these options are consistent with both an Alt-Svc field value and its
+HTTPS record.  This specification does not alter or clarify the
+interpretation of Alt-Svc's `protocol-id` field.
+
+Origins that publish an "ech" SvcParam in their HTTPS record SHOULD
+also publish an "ech" SvcParam for any alt-authorities.  Otherwise,
+clients might reveal the name of the server in an unencrypted ClientHello.
+Similar consistency considerations could apply to future SvcParamKeys, so
+alt-authorities SHOULD carry the same SvcParams as the origin unless
+a deviation is specifically known to be safe.  Clients MAY impose their
+own consistency conditions on Alt-Svc connections, e.g. requiring ECH if the
+origin's SvcParams offered ECH.
```

Comment:
This sentence is unclear to me. Is the idea that the client would still fetch the origin's HTTPS record? Or is it that the client maintains some state from the last time it fetched the HTTPS record, which may have been quite some time ago? I assume it's not the former. But the latter is very stateful and probably deserves more specific text if that's the intent.

Reply:
We don't have consensus to do any specific version of this, so there's no need to write it out in great detail.  I've made this vaguer, which I hope is clearer: "Clients MAY impose their own consistency conditions on Alt-Svc connections, e.g. requiring ECH **if the origin is known to support** ECH.".

I wouldn't recommend Alt-Svc ECH consistency enforcement to a client implementor, but I think it's permissible within the very lax rules of Alt-Svc, and some people clearly want to implement it.  A correct and reliable implementation would have to re-query the origin's HTTPS record along with the alt-authority's in order to have them both fresh for a comparison (your first option).  However, a client that only wants partial consistency enforcement could gate this check on a cached hint (your second option), or apply the enforcement only when a connection to the origin is still open, or when the origin's HTTPS record is still in the local cache.

But this is all complicated and speculative, and I don't think it goes in the draft.

@martinthomson Please review this version.

Reply:
It is still my opinion that it would be very difficult to get any more specific on how to enforce matching origin ECH usage without getting way over-specified.

We shouldn't require fresh querying the origin HTTPS because the performance cost will be unacceptable to many clients.  And anything involving caching ECH state of previous connections or looking at what's going on in parallel connections from the same client would involve a lot of finicky details where the right thing to do really depends on what the client wants to do and details of how the client is implemented.

I think it is the right approach to ensure clients are capable of adding such enforcement if desired without mandating a specific approach, and I think the current text handles this well.

(And the overall main case is that an ECH-supporting server should have ECH config on all endpoints, origin and Alt-Svc, and then it is very clearly defined that ECH-supporting clients will use ECH.  All this other discussion is just regarding what to do in the very discouraged case where the domains have a very silly configuration of only offering ECH on some endpoints.)

Reply:
Oh, my comment was more general: I wasn't sure what the text was trying to say in the first place. Whatever prescriptiveness we want here, the text should still convey that clearly! Intentionally trying not to be prescriptive isn't a reason to be unclear. :-)

I see the new text says something slightly looser. I read this to mean "if you have some out-of-band information that some origin supports ECH, you can go act on that". Is that the intent here, or was this meant to capture some other scenario?

Reply:
I'm not sure what you mean by "out of band", but that sounds basically right.  If you know the origin is FOO, and the alt-authority is not FOO, you MAY decide not to use the Alt-Svc, for any value of FOO.  How you know that the origin is FOO is not specified here, but presumably involves retrieving the origin's HTTPS record at some point.

In my view, this instruction is normatively null, because clients can always decide not to use the Alt-Svc: RFC 7838 says "By their nature, alternative services are OPTIONAL".  However, several comments on this PR have asked for text on client enforcement of ECH consistency, so this text serves as a reminder that it is allowed.

Reply:
I wonder if maybe "impose their own consistency conditions" implies a specific reason/algorithm for doing this, when what we're really trying to do is clarify that clients can choose not to use any Alt-Svc names or any HTTPS endpoints within them for really whatever reason the client wants.  Would this be clearer: "Clients MAY impose their own consistency conditions on Alt-Svc connections, e.g. ..." -> "Clients MAY choose to disallow connections to any HTTPS endpoints, e.g. disallow endpoints without an "ech" SvcParam if the client is aware that the origin's HTTPS endpoints support ECH."

Reply:
OK, changed to

> As noted in Section 2.4 of
> {{AltSvc}}, clients MAY disallow any Alt-Svc connection according to their
> own criteria, e.g. disallowing Alt-Svc connections that lack ECH support
> if ECH is supported on the origin.

Reply:
I think this is the least clear of the three so far. :-P How is the client supposed to know that ECH is supported on the origin, when Alt-Svc causes them not to query the origin?

It's especially problematic because the "ech" consistency requirement is merely a SHOULD rather than a MUST. At the end of the day, we need to define enough of what this means so that we can clearly say what is and isn't _required_ of servers to deploy ECH. The combination of merely making it a SHOULD and this clairvoyant client MAY means it's quite plausible to read this to mean that servers needn't worry about consistency since the client could just repair it.

Reply:
> I think this is the least clear of the three so far. :-P How is the client supposed to know that ECH is supported on the origin, when Alt-Svc causes them not to query the origin?

I would do it either by adding more data to the Alt-Svc info cache to specify that the original connection (that provided the Alt-Svc header) used ECH, or I'd query the QUIC connection pool for a still-active connection to the origin and leave info there about whether or not ECH was used.  But I imagine this would all be very dependent on the client-specifics of how Alt-Svc info is stored and how other connections are kept track of, so I don't know if there's a reasonable way to get any more specific in the draft.  Overall, I assume any client that wanted to do something like this could find a way.

> It's especially problematic because the "ech" consistency requirement is merely a SHOULD rather than a MUST. At the end of the day, we need to define enough of what this means so that we can clearly say what is and isn't required of servers to correctly deploy ECH. The combination of merely making it a SHOULD and this clairvoyant client MAY means it's quite plausible to read this to mean that servers needn't worry about consistency since the client could just repair it.

The server requirements are only a SHOULD because servers are not required to support ECH and not required to have a reasonable configuration to ensure clients use ECH.

But if a server wants to do the right thing and have a good configuration, I think all the text is already pretty clear how to do that (by including "ech=" on all endpoints in the origin and Alt-Svc).  Section 9.2 states pretty clearly that an RRSet with mixed ECH configuration is vulnerable to downgrade attacks.  And the current draft text here says servers with ECH on the origin should also include it on Alt services or the client might make non-ECH-protected connections.

I think it would be getting redundant, but do you think we need a final clarifying "servers SHOULD NOT rely on clients doing this to avoid leaking unencrypted Client Hello" after the "clients MAY..."?

Reply:
Well, I guess this goes back to what this text is trying to say in the first place. Again, the problem isn't the prescriptiveness but clarity. If the intent is about adding bits to the Alt-Svc cache, or looking at live connections (which are not guaranteed to be present), that doesn't check if "ECH is supported on the origin" but "if the origin was previously known to support ECH". If that's the intent (I'm still not sure, since the text is unclear), something like that may be better.

I don't think not being required to support ECH has any bearing on SHOULD/MUST. The text already only applies to "Origins that publish an 'ech' SvcParam in their HTTPS record". But, sure, a MUST may be impractical for other reasons. I'm not saying that it should be a MUST, but that the combination of the SHOULD and the client mitigation coming immediately afterwards in the same paragraph is misleading.

Perhaps simply remove it from the server paragraph, and instead integrate it into the paragraph about the client's rules?

Reply:
> if the origin was previously known to support ECH

I guess that's also slightly tricky because "previously" encompasses some random stale fetch you did 5 months ago, even though you have more recent information. "if the most recent fetch of the origin's HTTPS record supported ECH".

Though, thinking about this some more, maybe we're better off just dropping this. If the problem is a stale Alt-Svc record, the next time you visit the origin, when you learn about ECH support, you'll also get a new Alt-Svc record:

>   When an Alt-Svc response header field is received from an origin, its
>   value invalidates and replaces all cached alternative services for
>   that origin.

If the problem is a server misconfiguration, that basically the same as this one, which we haven't suggested clients repair:

> An HTTPS RRSet containing some RRs with "ech" and some without is vulnerable to a downgrade attack. This configuration is NOT RECOMMENDED.

Removing the text also doesn't change whether it's allowed because:

> As noted in Section 2.4 of {{AltSvc}}, clients MAY disallow any Alt-Svc connection according to their own criteria

What we're doing by calling it out is emphasizing it. That is, we actually want to suggest clients do this. But given how unwilling we are to write down what "this" is, and the inconsistency with the other analogous error, it's unclear to me that's actually true.

Finally, on SHOULD vs. MUST, I'll note that we already have other text where we have MUST-level requirements to capture things that, if you fail to do, things will go wrong.

> When publishing a record containing an "ech" parameter, the publisher MUST ensure that all IP addresses of TargetName correspond to servers that have access to the corresponding private key or are authoritative for the public name.

Reply:
I broke this sentence into its own paragraph and changed the example to

> e.g. disallowing Alt-Svc connections that lack ECH support when there is an active ECH-protected connection for this origin.

That's a specific example of an allowed, plausible behavior, and I don't think it will give any server operator the impression that they can safely skip ECH on their alt-authority.